packages/diagram-ui/src/features/viewport/touch.ts

```diff
@@ -0,0 +1,237 @@
+
+        const scaleChange = newDistance / this.lastTouchDistance!;
+        const newZoom = limit(viewport.zoom * scaleChange, this.viewerOptions.zoomLimits);
+
+        const dx = (newMidpoint.x - this.lastTouchMidpoint!.x) / viewport.zoom;
+        const dy = (newMidpoint.y - this.lastTouchMidpoint!.y) / viewport.zoom;
+        const offsetFactor = 1.0 / newZoom - 1.0 / viewport.zoom;
+        const newViewport = {
+            scroll: {
+                x: viewport.scroll.x - dx - offsetFactor * newMidpoint.x,
+                y: viewport.scroll.y - dy - offsetFactor * newMidpoint.y
+            },
+            zoom: newZoom
+        };
+
+        this.lastTouchDistance = newDistance;
+        this.lastTouchMidpoint = newMidpoint;
+        return newViewport;
+    }
+
+    /**
+     * Handler for the touch end event
+     *
+     * @param target the target element
+     * @param event the touch event
+     * @returns an array of actions to be dispatched
+     */
+    private touchEnd(target: SModelElementImpl, event: TouchEvent): Action[] {
+        if (event.touches.length === 0) {
+            this.lastTouchPosition = undefined;
+            this.lastTouchDistance = undefined;
+            this.lastTouchMidpoint = undefined;
+        } else if (event.touches.length === 1) {
+            this.lastTouchPosition = {
+                x: event.touches[0].clientX,
+                y: event.touches[0].clientY
+            };
+        }
+        return [];
+    }
+
+    /**
+     * Calculates the distance of the firsts two touches in the list
+     *
+     * @param touches the list of touches
+     * @returns the distance between the first two touches
+     */
+    private calculateDistance(touches: TouchList): number {
+        const dx = touches[0].clientX - touches[1].clientX;
+        const dy = touches[0].clientY - touches[1].clientY;
+        return Math.sqrt(dx * dx + dy * dy);
+    }
+
+    /**
+     * Returns the midpoint of the first two touches in the list relative to the canvas bounds
+     *
+     * @param touches the list of touches
+     * @param canvasBounds the bounds of the canvas in page coordinates
+     * @returns the midpoint of the first two touches
+     */
+    private calculateMidpoint(touches: TouchList, canvasBounds: Bounds): Point {
+        const windowScroll = getWindowScroll();
+        return {
+            x: (touches[0].clientX + touches[1].clientX) / 2 + windowScroll.x - canvasBounds.x,
+            y: (touches[0].clientY + touches[1].clientY) / 2 + windowScroll.y - canvasBounds.y
+        };
+    }
+
+    /**
+     * Handles a touch event by delegating to the appropriate handler
+     *
+     * @param methodName the name of the handler method
+     * @param model the model element
+     * @param event the touch event
+     */
+    private handleEvent(methodName: "touchStart" | "touchMove" | "touchEnd", model: SModelRootImpl, event: TouchEvent) {
+        const actions = this[methodName](model, event);
```

Comment:
I see what you are doing here.
But doesn't JS support first class functions?
Why not pass the corresponding method directly instead of doing a lookup?

Reply:
1. no bind necessary
2. because sprotty does it like this (and this may or may not be evolved into a PR in the future)